test/replication-luatest/gh_5158_invalid_recovery_from_xlog_test.lua

```diff
@@ -0,0 +1,78 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local fio = require('fio')
+local log = require('log')
+
+local g = t.group('gh_5158')
+
+g.before_each(function(cg)
+    cg.cluster = cluster:new({})
+    local box_cfg = {
+        checkpoint_count = 5,
+        log_level = 6,
+    }
+
+    cg.master = cg.cluster:build_server({alias = 'master', box_cfg = box_cfg})
+
+    box_cfg = {
+        replication = server.build_instance_uri('master'),
+    }
+
+    cg.replica = cg.cluster:build_server({alias = 'replica', box_cfg = box_cfg})
+    pcall(log.cfg, {level = 6})
+    cg.cluster:add_server(cg.master)
+    cg.cluster:add_server(cg.replica)
+    cg.cluster:start()
+end)
+
+g.after_each(function(cg)
+    cg.cluster:drop()
+end)
+
+g.test_invalid_recovery_from_xlog = function(cg)
+    cg.replica:stop()
+    t.helpers.retrying({}, function()
+        t.assert_str_matches(cg.master:exec(function()
+            return box.info.replication[2].downstream["status"]
+        end), 'stopped')
+    end)
+
+    cg.master:exec(function()
+        box.schema.space.create("test")
+        box.snapshot()
+        box.space.test:create_index("pk")
+        box.snapshot()
+    end)
+
+    cg.replica:start()
```

Comment:
Why did you stop the replica initially? I don't see why this is needed.

Reply:
I still don't understand why these stop and start are needed. I removed that part and the test worked just fine. It passed every time with the patch and failed every time without it:
```diff
diff --git a/test/replication-luatest/gh_5158_invalid_recovery_from_xlog_test.lua b/test/replication-luatest/gh_5158_invalid_recovery_from_xlog_test.lua
index c3f562ded..e50df5f75 100644
--- a/test/replication-luatest/gh_5158_invalid_recovery_from_xlog_test.lua
+++ b/test/replication-luatest/gh_5158_invalid_recovery_from_xlog_test.lua
@@ -29,13 +29,6 @@ g.after_each(function(cg)
 end)
 
 g.test_invalid_recovery_from_xlog = function(cg)
-    cg.replica:stop()
-    t.helpers.retrying({}, function()
-        t.assert_str_matches(cg.master:exec(function()
-            return box.info.replication[2].downstream["status"]
-        end), 'stopped')
-    end)
-
     cg.master:exec(function()
         box.schema.space.create("test")
         box.snapshot()
@@ -43,8 +36,6 @@ g.test_invalid_recovery_from_xlog = function(cg)
         box.snapshot()
     end)
 
-    cg.replica:start()
-
     -- Wait until everything is replicated from the master to the replica
     t.helpers.retrying({}, function()
         cg.replica:wait_vclock_of(cg.master)
```